docs/src/pages/nextjs/appdir.mdx

```diff
@@ -119,4 +119,54 @@ export default function Home() {
 }
 ```
 
+### Configure automatic hydration during SSR (recommended)
+
+> Added in `v5.6`
+
+By default, `useUploadThing` (and thus also the premade components) will fetch
+your server for the route endpoint permissions which will show a loading state.
+By utilizing server components, we can perform automatic hydration of this data
+instead to avoid the loading state and show the route permissions immediately.
+
+To add automatic SSR hydration, simply render the `<NextSSRPlugin />` hydration
+helper in the body of your root layout **before** the children. The before and
+after is shown below (refresh page to see the difference):
+
+import { WithoutSSR, WithSSR } from "../../components/ssr-diff";
+
+<div className="mt-6 flex items-center justify-center gap-8">
+  <WithoutSSR />
+  <WithSSR />
+</div>
+
+```tsx copy filename="app/layout.tsx"
+import { NextSSRPlugin } from "@uploadthing/react/next-ssr-plugin";
+import { extractRouterConfig } from "uploadthing/server";
+
+import { uploadRouter } from "~/server/uploadthing";
+
+export default function RootLayout({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  return (
+    <html lang="en">
+      <body>
+        <NextSSRPlugin
+          /**
+           * The `extractRouterConfig` will extract **only** the route configs
+           * from the router to prevent additional information from being
+           * leaked to the client. The data passed to the client is the same
+           * as if you were to fetch `/api/uploadthing` directly.
+           */
+          routerConfig={extractRouterConfig(uploadRouter)}
```

Comment:
I wish we could just pass the whole router and the component could do this extraction for the user.

Reply:
The component is a client component though so i think that will leak stuff we dont want to the client

Reply:
Yeah, hence why it remains a wish unfortunately :(

Reply:
I guess we could make a wrapper component but i think it's hard getting that to bundle split correctly:

```tsx
// a.ts
function SSRPlugin(props: { router: FileRouter }) {
  return <SSRPluginInner routerConfig={extractRouterConfig} />
}

// b.ts
"use client"
function SSRPluginInner(props: { routerConfig: ... }) {
  useServerInsertedHTML(...)
  
  return null
}
```

Reply:
Yeah it is not worth the trouble IMO, too likely to cause problems for not enough gain